src/auth/strategies/oidc.strategy.ts

```diff
@@ -131,34 +148,111 @@ export class OidcStrategy extends PassportStrategy(Strategy, "oidc") {
+            oidcUser[sourceField] = userinfo[targetField] as string;
+          } else if (Array.isArray(targetField) && targetField.length) {
+            const values = targetField
+              .filter((field) => field in userinfo)
+              .map((field) => userinfo[field] as string);
+
+            if (values.length) {
+              oidcUser[sourceField] = values.join("_");
+            }
+          }
+        },
+      );
+    }
+
     // Prior to OpenID Connect Basic Client Profile 1.0 - draft 22, the "sub"
     // claim was named "user_id".  Many providers still use the old name, so
-    // fallback to that.
-    const userId = userinfo.sub || (userinfo.user_id as string);
-    if (!userId) {
+    // fallback to that. https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+
+    if (!oidcUser.id) {
       throw new Error("Could not find sub or user_id in userinfo response");
     }
 
-    profile.id = userId;
-    profile.username = userinfo.preferred_username ?? userinfo.name ?? "";
-    profile.displayName = userinfo.name ?? "";
-    profile.emails = userinfo.email ? [{ value: userinfo.email }] : [];
-    profile.email = userinfo.email ?? "";
-    profile.thumbnailPhoto = this.getUserPhoto(userinfo);
+    profile.emails = oidcUser.email ? [{ value: oidcUser.email }] : [];
+    profile.thumbnailPhoto = this.getUserPhoto(oidcUser.thumbnailPhoto);
+    profile.oidcClaims = userinfo;
+
+    const oidcUserProfile = { ...oidcUser, ...profile };
+    return oidcUserProfile;
+  }
+
+  parseQueryFilter(userProfile: OidcProfile) {
+    const userQueryMappingFilter =
+      this.configService.get<IOidcUserQueryMapping>("oidc.userQueryMapping");
+
+    const operator = userQueryMappingFilter?.operator
+      ? "$" + userQueryMappingFilter?.operator.toLowerCase()
+      : undefined;
+
+    const filter = userQueryMappingFilter?.filter?.length
+      ? userQueryMappingFilter.filter.reduce(
+          (acc: Record<string, unknown>[], mapping: string) => {
+            if (!mapping.includes(":")) {
+              Logger.error(
+                "OIDC_USERQUERY_MAPPING_FILTER must follow the format 'name:name, email:email'",
+                mapping,
+              );
+            }
+            const [filterField, userProfileField] = mapping.split(":");
+            if (userProfileField in userProfile) {
+              acc.push({
+                [filterField]:
+                  userProfile[userProfileField as keyof UserProfile],
+              });
+            }
+            return acc;
+          },
+          [],
+        )
+      : undefined;
+
+    const userFilter: FilterQuery<UserDocument> =
+      !operator || !filter || filter.length < 1
+        ? {
+            $or: [
+              { username: userProfile.username },
+              { email: userProfile.email },
+            ],
+          }
+        : {
+            [operator]: filter,
+          };
 
-    return profile;
+    return userFilter;
   }
```

Comment:
I'm not sure I understand what this function does

Reply:
It's refactored and should be more readable now.
This is a function that transforms env variables into user query filter
`let user = await this.usersService.findOne(userFilter);`